1-Draft/RFC00XX-PredIctive-IntelliSense.md

```diff
@@ -0,0 +1,359 @@
+
+## Extension Model Specification
+
+### Predictive Suggestion Contract to Support Addition Providers
+
+Below is an abbreviated code listing of the API contract to support suggestions from additional
+providers.
+
+```csharp
+/// The class represents a predictive suggestion generated by a predictor.
+public sealed class PredictiveSuggestion
+{
+/// Gets the suggestion.
+public string SuggestionText { get; }
+
+/// Gets the tooltip of the suggestion.
+public string? ToolTip { get; }
+
+/// Initializes a new instance of the <see cref="PredictiveSuggestion"/> class.
+/// <param name="suggestion">The predictive suggestion text.</param>
+public PredictiveSuggestion(string suggestion)
+: this(suggestion, toolTip: null)
+```
+
+### Discovery of Provider Metadata and Registration
+
+Providers are registered automatically when the user installs a provider module. Removal of
+registration occurs when the module is removed. Users will need to discover installed providers
+along with additional metadata, such as version information to assist users in upgrading and
+diagnosing problems.
+
+To list the currently installed providers and their respective metadata:
+
+```powershell
+Get-SubSystem
+```
+
+### Expected Provider Performance
+
+Additional providers often will communicate with an internet connect service to provide suggestions.
+To minimize the appearance of slow responses, we estimate that the maximum roundtrip time of 20ms is
+permissible to not negatively impact users.
+
+If the roundtrip time is exceeded, the predictor model will provide historical suggestions, or
+suggestions from another provider that completed in the allowed time.
+
+* Default required provider performance = 20ms
+
+The current roundtrip time in PSReadLine using the Tab Completion feature is approximately **2-3ms**. We
+believe that this current default roundtrip time for additional providers may still exceed user
+expectations. With additional testing and feedback, we may need to reduce the roundtrip time to less
+than **20ms**.
+
+### Registration for Module Owners
+
+Predictive IntelliSense is an extensible model offering module authors the opportunity to create their own predictor plugins. Customers using Predictive IntelliSense will benefit from increased coverage of domain specific technologies and other helpful predictors.
+
+Module authors writing new predictors will need to register their predictor with the extension
+model. Predictor registration is accomplished with the **OnImport** and **OnRemove** methods
+executed on module startup and removal. Additional documentation is available for
+[IModuleAssemblyInitializer
+(OnImport)](https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.imoduleassemblyinitializer?view=powershellsdk-7.0.0)
+and
+[IModuleAssemblyCleanup (OnRemove)](https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.imoduleassemblycleanup?view=powershellsdk-7.0.0).
```

Comment:
Registration is done by calling the subsystem registration API. `OnImport` and `OnRemove` are the places to call the registration API for a binary module.